src/core/plugin/PluginController.cpp

```diff
@@ -1,54 +1,106 @@
+                if (!plugin || !plugin->isValid()) {
+                    g_warning("Error loading plugin \"%s\"", f.path().string().c_str());
+                    continue;
+                }
+                plugin->setEnabled(plugin->isDefaultEnabled());
+                returner.emplace_back(std::move(plugin));
+            } catch (std::exception const& e) {
+                g_warning("Error loading plugin \"%s\": %s", f.path().string().c_str(), e.what());
             }
+        }
+    } catch (fs::filesystem_error const& e) {
+        g_warning("Could not open plugin dir: \"%s\": %s", path.string().c_str(), e.what());
+    }
+    return returner;
+}
 
-            if (plugin->isDefaultEnabled()) {
-                plugin->setEnabled(!(std::find(pluginDisabled.begin(), pluginDisabled.end(), plugin->getName()) !=
-                                     pluginDisabled.end()));
-            } else {
-                plugin->setEnabled(std::find(pluginEnabled.begin(), pluginEnabled.end(), plugin->getName()) !=
-                                   pluginEnabled.end());
-            }
+auto emplace_sorted_if_not_exists(std::vector<std::unique_ptr<Plugin>>& plugins, std::unique_ptr<Plugin> plugin)
+        -> void {
+    auto const& path = plugin->getPath();
+    auto endit = std::end(plugins);
+    auto it = std::find_if(std::begin(plugins), endit,
+                           [&path](auto const& p) { return fs::canonical(p->getPath()) == fs::canonical(path); });
+    if (it != endit) {
+        return;
+    }
+    plugins.emplace_back(std::move(plugin));
+    std::sort(begin(plugins), end(plugins), [](auto const& lhs, auto const& rhs) {
+        auto tieer = [](auto&& p) { return std::tie(p->getName(), p->getPath()); };
+        return tieer(lhs) < tieer(rhs);
+    });
+}
+}  // namespace
+#endif
 
-            plugin->loadScript();
 
-            this->plugins.emplace_back(std::move(plugin));
+PluginController::PluginController(Control* control): control(control) {
+#ifdef ENABLE_PLUGINS
+    // Todo(fabian) move those search paths into PathUtils
+    auto searchPath = control->getGladeSearchPath()->getFirstSearchPath();
+    auto searchPaths = {fs::weakly_canonical(searchPath /= "../plugins"),  //
+                        fs::weakly_canonical(searchPath /= "/plugins"),    //
+                        Util::getConfigSubfolder("plugins")};
+
+    for (auto&& path: searchPaths) {
+        for (auto&& plugin: load_awailable_plugins_from(path, control)) {
+            emplace_sorted_if_not_exists(plugins, std::move(plugin));
+        }
+    }
+
+    Settings* settings = control->getSettings();
+    // enable plugins
+    std::vector<std::string> pluginEnabled = StringUtils::split(settings->getPluginEnabled(), ',');
+    for (auto&& plugin_path: pluginEnabled) {
+        if (auto iter = std::find_if(begin(plugins), end(plugins),
+                                     [&plugin_path](auto&& plugin) { return plugin->getPath() == plugin_path; });
```

Comment:
This is `O(len(enabled) * n) ~= O(n^2)` assuming `enabled = n`. If the map above is defined, then I think we can turn this into a map lookup.

Reply:
Or use `std::lower_bound` as noted above.